src/mc-fle2-payload-iev.c

```diff
@@ -151,36 +153,60 @@ mc_FLE2IndexedEncryptedValue_add_S_Key (_mongocrypt_crypto_t *crypto,
 
    /* Use TokenKey to create ServerDataEncryptionLevel1Token and decrypt
     * InnerEncrypted. */
-   {
-      mc_ServerDataEncryptionLevel1Token_t *token =
-         mc_ServerDataEncryptionLevel1Token_new (crypto, &TokenKey, status);
-      if (!token) {
-         return false;
-      }
 
-      const _mongocrypt_buffer_t *token_buf =
-         mc_ServerDataEncryptionLevel1Token_get (token);
-      uint32_t bytes_written;
+   mc_ServerDataEncryptionLevel1Token_t *token =
+      mc_ServerDataEncryptionLevel1Token_new (crypto, &TokenKey, status);
+   if (!token) {
+      return false;
+   }
 
-      const uint32_t plaintext_len = _mongocrypt_fle2_calculate_plaintext_len (
-         iev->InnerEncrypted.len, status);
-      if (!plaintext_len) {
-         mc_ServerDataEncryptionLevel1Token_destroy (token);
-         return false;
-      }
-      _mongocrypt_buffer_resize (&iev->Inner, plaintext_len);
-
-      /* Decrypt InnerEncrypted. */
-      if (!_mongocrypt_fle2_do_decryption (crypto,
-                                           token_buf,
-                                           &iev->InnerEncrypted,
-                                           &iev->Inner,
-                                           &bytes_written,
-                                           status)) {
-         mc_ServerDataEncryptionLevel1Token_destroy (token);
-         return false;
-      }
-      mc_ServerDataEncryptionLevel1Token_destroy (token);
+   bool ret = mc_FLE2IndexedEncryptedValue_decrypt(crypto, iev, token, NULL, NULL, status);
+   
+   mc_ServerDataEncryptionLevel1Token_destroy (token);
+
+   return ret;
+}
+
+static bool
+mc_FLE2IndexedEncryptedValue_decrypt (_mongocrypt_crypto_t *crypto,
+                                        mc_FLE2IndexedEncryptedValue_t *iev,
+                                        mc_ServerDataEncryptionLevel1Token_t* token,
+                                       bool(callback)(mc_reader_t*, void*, mongocrypt_status_t* status),
+                                        void* callback_data,
```

Comment:
```suggestion
                                       mc_FLE2IndexedEqualityEncryptedValueTokens* tokens /* optional */,
```

Instead of a callback, can `tokens` be an optional out parameter?

Reply:
I prefer it here since there is only one caller that sets the callback.